src/Compilers/CSharp/Portable/Symbols/ExtraAnnotations.cs

```diff
@@ -0,0 +1,178 @@
+                { "System.Void System.Buffer.BlockCopy(System.Array, System.Int32, System.Array, System.Int32, System.Int32)", Array(default, Nullable(false), default, Nullable(false), default, default) },
+                { "System.Int32 System.Buffer.ByteLength(System.Array)", Array(default, Nullable(false)) },
+                { "System.Byte System.Buffer.GetByte(System.Array, System.Int32)", Array(default, Nullable(false), default) },
+                { "System.Void System.Buffer.SetByte(System.Array, System.Int32, System.Byte)", Array(default, Nullable(false), default, default) },
+                { "System.Byte System.Byte.Parse(System.String)", Array(default, Nullable(false)) },
+                { "System.Byte System.Byte.Parse(System.String, System.IFormatProvider)", Array(default, Nullable(false), default) },
+                { "System.Byte System.Byte.Parse(System.String, System.Globalization.NumberStyles)", Array(default, Nullable(false), Nullable(false)) },
+                { "System.Byte System.Byte.Parse(System.String, System.Globalization.NumberStyles, System.IFormatProvider)", Array(default, Nullable(false), Nullable(false), default) },
+                { "System.String System.String.Concat(System.String, System.String)", Array(Nullable(false), Nullable(true), Nullable(true)) },
+            }.ToImmutableDictionary();
+
+        private static readonly ImmutableDictionary<string, ImmutableArray<ImmutableArray<AttributeDescription>>> Attributes =
+            new Dictionary<string, ImmutableArray<ImmutableArray<AttributeDescription>>>
+            {
+                { "System.Boolean System.String.IsNullOrEmpty(System.String)", Array(default, NotNullWhenFalse) },
+                { "System.Boolean System.String.IsNullOrWhiteSpace(System.String)", Array(default, NotNullWhenFalse) },
+                { "System.Boolean System.String.Contains(System.String)", Array(default, EnsuresNotNull) },
+            }.ToImmutableDictionary();
+
+        internal static string MakeMethodKey(PEMethodSymbol method, ParamInfo<TypeSymbol>[] paramInfo)
+        {
+            var pooledBuilder = PooledStringBuilder.GetInstance();
+
+            StringBuilder builder = pooledBuilder.Builder;
+            Add(paramInfo[0].Type, builder);
+            builder.Append(' ');
+
+            Add(method.ContainingType, builder);
+            builder.Append('.');
+
+            builder.Append(method.Name);
+            builder.Append('(');
+
+            for (int i = 1; i < paramInfo.Length; i++)
+            {
+                Add(paramInfo[i].Type, builder);
+                if (i < paramInfo.Length - 1)
+                {
+                    builder.Append(", ");
+                }
+            }
+
+            builder.Append(')');
+            return pooledBuilder.ToStringAndFree();
+
+            // PROTOTYPE(NullableReferenceTypes): Many cases are not yet handled
+            // generic type args
+            // ref kind
+            // 'this'
```

Comment:
>'this' [](start = 15, length = 6)

Does this represent an extension method `this` parameter? Is that important?

Reply:
I'm not sure if that'll be important. I just wanted to keep some notes of things the test if/when we generalize the external annotations.